### ISSUE TITLE
[Dataloader] - Invalid reuse of previous injected db_context or data-loader itself..

### ISSUE THREAD

**Issue by damikun** [ChilliCream/hotchocolate#3690]
**Describe the bug**

Notes:
- This is race-condition and appears only sporadically...
- The DB context can be Transient or pooled and provides a new instance on every inject...
- **This means data-loader body is not thread-save**... DI objects injected (does not meter what is lifetime) because the instance of data-loader is shared and parallel task reuse it... You can have pooled/transient context but body of loader is shared between threads. This can be well tested including semaphore with counter inside loader.....

HC data-loader uses the previous db context. or loader itself is reused before prev. finished the job...

An example query is resolving  `project_assigned_users` > `user` and fails as `A second operation started on context...` but as I wrote DB context should be unique from .net core Setup... And this race condition appears only with dataloaders let's say 2-3x in 50 requests... 

```
projects {
 edges {
  node {
    assignedUsers{
     edges {
      node {
       ... user_fragment
  }}}}}
```

![image](https://user-images.githubusercontent.com/54806834/118122021-9056d780-b3f2-11eb-9a85-bdf650497a77.png)

Fails after hits common user dataloader which is used across the entire app... But only appears as a problem if resolving is deep nested.. Like on example 2x nested connections... 

```
public async Task<GQL_User> GetProjectUser(
       GQL_ProjectUser user,
       UserByIdDataLoader loader,
       CancellationToken cancellationToken) {

       if (user != null && user.UserID != 0) {
            return await loader.LoadAsync(user.UserID, cancellationToken);
       }
       return null;
}

```
**Version:**
- Net5
- EF Core 5
- HC 11.2.2

**Comment by ronaldvdv** [comment]
I feel I may see a different but related issue - when I'm using a dataloader at multiple places in a nested query, I'm seeing a "Cannot access a disposed object" error for the `DbContext`.

Feels like what @damikun is saying - _because the instance of data-loader is shared and parallel task reuse it_

**Comment by damikun** [comment]
It can be equal problem but also make sure your DB context is registered well and support concurrent state... Make sure it is not request scoped... and each injection gives you unique object or pooled one..

The temporary workaround is to wrap body of data-laoder with SemaphoreSlim as quick fix...

**Comment by ronaldvdv** [comment]
> It can be equal problem but also make sure your DB context is registered well and support concurrent state... Make sure it is not request scoped... and each injection gives you unique object or pooled one..
> 
> The temporary workaround is to wrap body of data-laoder with SemaphoreSlim as quick fix...

In my case I'm using pooled dbcontexts; what should the db context registration look like? Also I guess using a SemaphoreSlim could solve a `A second operation...` exception but probably not a `Cannot access a disposed object` exception.

**Comment by ocecaco** [comment]
I have also run into this issue. The root cause of this issue seems to be [this function](https://github.com/ChilliCream/hotchocolate/blob/87a887d14c4a025c75f8a5db886053ac83ca48a9/src/GreenDonut/src/Core/DataLoaderBase.cs#L316). That function is responsible for creating a new batch once the previous batch has either reached its maximum size or is being executed already (and therefore can no longer be extended with additional keys). Note that the function simply replaces the current batch with a new batch, and no attempt is made to ensure that the current batch has been completed before scheduling the batch query function of the new batch.

To fix this, it seems like it would either be necessary to have the data loader wait until the previous batch has been completed before submitting a new batch, or to use a batch scheduler that does not process multiple batch items concurrently (so that even if you submit a new batch to the scheduler before the previous one has been completed, it won't get executed yet by the scheduler). Alternatively, the client could provide their own synchronization as mentioned in one of the comments above, but it seems painful to require clients to handle thread safety by default, so it would at least be nice to have a `ThreadSafeDataLoader` included in the library.

This race condition does not usually manifest in my code base, unless I am using stacked DataLoaders (i.e. `await LoadAsync` called multiple times on different data loaders in the same resolver). I suspect that this is because without stacked DataLoaders, all of my keys just happen to end up in the same batch, whereas with stacked DataLoaders you get "partial" batches because some of the resolvers are still at the first `await` and reach the second `await` while the first "partial" batch is executing (triggering the race condition).

**Comment by damikun** [comment]
**As workaround I can recommend you to:**
- Use pooled db context factory - each time you request new object from pool
- Wrap body using SemaphoreSlimto lock body... Its not big problem since this problem is not with every request and is more sporadic..

**Comment by ocecaco** [comment]
@damikun Thanks, it seems like that is the way to go for now. Although I think it might just be a coincidence (or at least an undocumented guarantee that cannot be relied upon) that the race condition does not usually occur in most DataLoader calls. So I would think that to be absolutely correct you would have to wrap _every_ DataLoader that uses a database connection (i.e. in my code base every DataLoader) in a semaphore. At least that is true unless there is some implementation guarantee like "we guarantee that DataLoader will not be called concurrently unless you are calling multiple DataLoaders in a single resolver".

I guess it might not be that much of an issue in practice if the race condition rarely triggers, because you could look for the DataLoaders that are causing issues and add the semaphore there (and pray that the other ones don't cause errors in production too frequently).

**Comment by darren-clark** [comment]
This is an old ticket, but I ran across this as well. The issue is that the same data loader instance can be called multiple times in parallel, depending on when earlier tasks complete so constructor injected EF context ends up being reused. In my case I had two queries returning a parent object, one returning 1 and one returning 150. Sometimes the loader would get called to resolve the field for both sets, and sometimes once per set.  My fix for this was to inject IDbContext factory instead.
